controller/src/scheduler.rs

```diff
@@ -0,0 +1,387 @@
+    Windowed,
+    Oneshot,
+}
+
+fn determine_schedule_type(schedule: &Schedule) -> Result<ScheduleType> {
+    let duration_between_each_schedule_datetime = duration_between_next_two_points(&schedule)?;
+    Ok(
+        if duration_between_each_schedule_datetime.num_seconds() == 1 {
+            ScheduleType::Windowed
+        } else {
+            ScheduleType::Oneshot
+        },
+    )
+}
+
+fn duration_between_next_two_points(schedule: &Schedule) -> Result<chrono::Duration> {
+    let first_time = schedule
+        .upcoming(Utc)
+        .next()
+        .context(scheduler_error::GetScheduledDatetimeSnafu)?;
+    let second_time = schedule
+        .upcoming(Utc)
+        .nth(1)
+        .context(scheduler_error::GetScheduledDatetimeSnafu)?;
+    Ok(second_time - first_time)
+}
+
+fn std_duration(d: &chrono::Duration) -> Result<std::time::Duration> {
+    Ok(d.to_std()
+        .context(scheduler_error::ConvertToStdDurationSnafu)?)
+}
+
+fn determine_update_window_or_scheduler(
+    update_window_start: &str,
+    update_window_stop: &str,
+    scheduler: &str,
+) -> String {
+    let scheduler_cron_expression = if scheduler != SCHEDULER_DEFAULT {
+        scheduler.to_string()
+    } else {
+        if update_window_start != UPDATE_WINDOW_START_DEFAULT.to_string()
+            || update_window_stop != UPDATE_WINDOW_STOP_DEFAULT.to_string()
+        {
+            //convert update time window values to cron expression.
+            cron_expression_converter(update_window_start.clone(), update_window_stop.clone())
+        } else {
+            scheduler.to_string()
+        }
+    };
+
+    scheduler_cron_expression
+}
+
+/// Convert update time window values to cron expression
+/// For overnight schedule, we format it to `{}-23,0-{}` on hour spot.
+/// For example, start time on 18 and stop time on 5. we format it to `18-23,0-5`
+fn cron_expression_converter(update_window_start: &str, update_window_stop: &str) -> String {
+    let start_hour: u8 = update_window_start
+        .split(":")
+        .next()
+        .unwrap()
+        .parse()
+        .unwrap();
+    let stop_hour: u8 = update_window_stop
+        .split(":")
+        .next()
+        .unwrap()
+        .parse()
+        .unwrap();
+
+    // if update window envs are both on default(0:0:0), it should return "* * * * * * *"
+    if update_window_start == UPDATE_WINDOW_START_DEFAULT.to_string()
+        && update_window_stop == UPDATE_WINDOW_STOP_DEFAULT.to_string()
+    {
+        return SCHEDULER_DEFAULT.to_string();
+    }
+
+    let cron_expression_hour = if start_hour <= stop_hour {
+        format!("* * {}-{} * * * *", start_hour, stop_hour)
+    } else {
+        format!("* * {}-23,0-{} * * * *", start_hour, stop_hour)
+    };
+
+    cron_expression_hour
+}
+
+#[cfg(test)]
+pub(crate) mod test {
+    use super::*;
+    use chrono::{NaiveDate, Utc};
+
+    #[tokio::test]
+    async fn test_duration_between_next_two_points() {
+        let test_cases = vec![
```

Comment:
Are you sure these tests are deterministic? The reason I suggested passing in a DateTime<Utc> to the duration_between_next_to_points function and using .after instead of .upcoming was so that the function would be deterministic for testing

Reply:
I think those tests are deterministic since I didn't specify a cron expression which could be expired. But yeah, I think overall `after` is a good idea than `upcoming` during test.